app/controllers/sessions_controller.rb

```diff
@@ -2,54 +2,84 @@ class SessionsController < ApplicationController
   def new
   end
 
+  def shop_owner_login
+  end
+
   def create
     if request.env["omniauth.auth"]
       oauth_user = OauthUser.new(request.env["omniauth.auth"])
       user = oauth_user.login_or_create
       session[:user_id] = user.id
 
       return redirect_to_user_intended unless session[:user_intended].nil?
-      redirect_to root_path, notice: "Welcome back, #{current_user.first_name}"
+      redirect_to root_path, notice: welcome_user
     else
-      process_form_login
+      process_form_login "RegularUser"
     end
   end
 
+  def shop_owner_create
+    process_form_login "ShopOwner"
+  end
+
   def destroy
-    @current_user = nil
-    session.delete :user_id
-    redirect_to root_path, notice: "Logged out Successfully!"
+    logout_user current_user
+  end
+
+  def shop_owner_destroy
+    logout_user current_shop_owner
   end
 
   private
 
-  def process_form_login
-    user = RegularUser.find_by_email(params[:session][:email])
-    return if user_is_nil?(user)
-    if user.authenticate(params[:session][:password]) && user.active
-      login_successful user
+  def logout_user(_user)
+    _user = nil
+    session.delete :user_id
+    redirect_to root_path, notice: MessageService.logout
+  end
+
+  def process_form_login(model)
+    user = model.constantize.find_by(email: params[:session][:email])
+    return if user_is_nil?(user, model)
+
+    authenticated = user.authenticate(params[:session][:password])
+    if authenticated && user.verified && user.active
+      login_successful(user, model)
     else
-      password_invalid
+      password_invalid model
     end
   end
 
-  def user_is_nil?(user)
+  def user_is_nil?(user, model)
```

Comment:
`model` is `RegularUser` or `ShopOwner` and it is used to redirect to the login form page or shop_owner_login form page respectively.